tests/binary_agreement_mitm.rs

```diff
@@ -0,0 +1,210 @@
+                        self.stage = match self.stage {
+                            ReorderingStage::DeliverBBVal => ReorderingStage::DeliverA0BVal,
+                            ReorderingStage::DeliverA0BVal => ReorderingStage::DeliverA0Aux,
+                            ReorderingStage::DeliverA0Aux => ReorderingStage::Finished,
+                            ReorderingStage::Finished => unreachable!(),
+                        };
+                        self.stage_progress = 0;
+                    }
+                }
+            }
+        }
+    }
+
+    fn tamper(
+        &mut self,
+        mut net: NetMutHandle<BinaryAgreement<NodeId>>,
+        _: NetMessage<BinaryAgreement<NodeId>>,
+    ) -> Result<Step<BinaryAgreement<NodeId>>, CrankError<BinaryAgreement<NodeId>>> {
+        const BVAL_FALSE_MSG: Message = Message {
+            epoch: 0,
+            content: MessageContent::SbvBroadcast(SbvMessage::BVal(false)),
+        };
+        const BVAL_TRUE_MSG: Message = Message {
+            epoch: 0,
+            content: MessageContent::SbvBroadcast(SbvMessage::BVal(true)),
+        };
+        if !self.sent_initial_messages {
+            self.sent_initial_messages = true;
+            for target in 1..(1 + NODES_PER_GROUP) {
+                // Disagree with the nodes in A0.
+                net.inject_message(
+                    QueuePosition::Front,
+                    NetMessage::<BinaryAgreement<NodeId>>::new(0, BVAL_TRUE_MSG.clone(), target),
+                );
+            }
+            for target in (1 + NODES_PER_GROUP)..(1 + NODES_PER_GROUP * 2) {
+                // Agree with the nodes in A1.
+                net.inject_message(
+                    QueuePosition::Front,
+                    NetMessage::<BinaryAgreement<NodeId>>::new(0, BVAL_FALSE_MSG.clone(), target),
+                );
+            }
+        } else if self.stage == ReorderingStage::Finished && !self.sent_final_messages {
+            self.sent_final_messages = true;
+            for target in 1..(1 + NODES_PER_GROUP * 2) {
+                // Both agree and disgree with nodes in A.
+                net.inject_message(
+                    QueuePosition::Front,
+                    NetMessage::<BinaryAgreement<NodeId>>::new(0, BVAL_FALSE_MSG.clone(), target),
+                );
+                net.inject_message(
+                    QueuePosition::Front,
+                    NetMessage::<BinaryAgreement<NodeId>>::new(0, BVAL_TRUE_MSG.clone(), target),
+                );
+            }
+        }
+        Ok(Step::default())
+    }
+}
+
+#[test]
+fn reordering_attack() {
+    let _ = env_logger::try_init();
+    let ids: Vec<NodeId> = (0..NUM_NODES).collect();
+    let mut net = NetBuilder::new(ids.iter().cloned())
+        .adversary(ReorderingAdversary::default())
+        .crank_limit(10000)
+        .using(|info| {
+            BinaryAgreement::new(Arc::new(info.netinfo), 0, info.id)
```

Comment:
The third argument ("`proposer_id`") needs to be the same in all nodes.

That's indeed a confusing API, of course, and we should definitely change it. But for now, you can just replace `info.id` with `0`.